### DISCUSSION TITLE
[BUG] Challenge times increase during final copy of plots

### DISCUSSION THREAD

**Discussion by Akaino** [Chia-Network/chia-blockchain#3220] [Support]
**Might not be an actual bug**
Challenges take longer when new plots are being copied to any plot directory.

**To Reproduce**
Plot something. Wait for final copy. Check challenge times.

**Expected behavior**
Challenge times shouldn't take longer, I guess?

**Screenshots**
Challenges during copy
![Screenshot1](https://i.imgur.com/gv1C8sa.png)

Plot 1/2 copied successfully
![Screenshot2](https://i.imgur.com/kV6i5h6.png)

Plot 2/2 copied successfully
![Screenshot3](https://i.imgur.com/FcAY7KH.png)

**Desktop (please complete the following information):**

- OS: Windows 10 [10.0.19041]
- OS Version/Flavor: Windows 10 [10.0.19041]
- CPU: Intel Core i7 9900k

**Additional context**
The application simply reads logs from the default logs directory and throws data into the graph and table.

Plotting with cli, 2 parallel, staggered by 60 minutes:
> chia plots create -k 32 -b 4000 -r 4 -u 128 -n 28 -t T:\Temp -2 T:\Temp2 -d Q:\
>>T: = TempDir [NVME]
Q; = FinalDir [SATA HDD] or [Synology NAS] (issue exists in both cases).

So far I'm not missing challenges.

**Comment by sargonas** [comment]
This is likely due to the fact that the challenge time is a result of system resource availability, especially drive I/O . If you are peaking some of your I/O on the drive you’re farming from when trying to copy the files into it, then the challenges are going to take a few milliseconds longer while the copy is underway as the threads wait their turn.

**Reply by Akaino** [reply]
This is not the case here though. I'm not talking about a few ms. Check the screenshots. I'm talking seconds here. It even reached >300s for a few challenges. As soon as the copy was over everything went back to normal.

Also, I don't get why issues are being closed after **_assuming_** why that bug happens. Even without coming up with a possible solution.

**Reply by sargonas** [reply]
Apologies, I read your screen shots wrong (lots of tiny text, small window, you know how it is) and only looked at the compared red boxes in the last two. I see now that yes, you are getting 1-8 second delays there, and I even see a rogue 18 hidden in the midst.

What is your drive setup, as in, what kind of drive are you copying over for the -2, type, connection method, etc? (SATA, NAS< USB, (3.0, -C, etc)

Apologies for closing it, but this has come up many times and historically ends up being a slow I/O bottleneck at that stage or config issue, so I overzealously closed this one when responding. You post is still open but it was moved over to the support section of discussions, and as it's noodled though the problem, if anyone determines it is indeed a bug in the codebase then someone will spin up a bug tracking issue off of it.

**Reply by Akaino** [reply]
Sure, no worries. Nice to see you responding here!
The setup is quite simple.

Windows PC farming via PowerShell with the given commands. -t and -2 both target an internal m.2 nvme. -d however targets either internal SATA HDD (6-15s here) or Synology ds920+ (NAS) over 1Gbit Network (30-50s here).

I understand that the NAS copy takes some system resources. I don't really get why it drives the challenge times down THAT much. 

I don't think it's a network issue. NAS only takes ~120MB/s as it's connected via single Ethernet. Also, i7 9900k should be able to handle the copy load and also do the challenges.
There's always at least 2 threads free for everythin OS related.
Memory never goes above 28000.

**Reply by sargonas** [reply]
aha! im curious, if you take the Synology out of rotation for a bit and only run the SATA farm drives, what do your times look like? I've seen tons of discussion in the Keybase chat that NAS's, including Synology, just give all sorts of unreliable timings. A high end Synology SHOULD be fine, but the community consensus seems to be to avoid them. I'm not saying it for sure is yet, but an A|B test without it would help narrow things down.

**Reply by Akaino** [reply]
I'll give it a shot and drop the plottings targeting the NAS as final.

**Reply by Akaino** [reply]
Ok, I figured copying to the NAS indeed does increase challenge times **a lot**.
I'm constantly getting challenges >20s with a lot >30s. Even some 400-500s!!
However, I don't get why that happens. I understand that a NAS might not be the perfect storage considering network traffic etc. But copying to it really shouldn't affect challenge times by **that** amount.
I'm going to try limiting bandwidth to the NAS to see whether it's enough to keep challenge times in check.

Edit:
I tried the following now:
Limiting Incoming Newtork Bandwidth on the NAS to ~0.5GBit. No difference.
Limiting Data Transfer with Windows QoS Group Policy. Policy was ignored (Did I say I hate Windows?).
Limiting SMB Data Transfer with PowerShell `New-NetQosPolicy -Name "SMBRestrictFileCopySpeed" -SMB -ThrottleRateActionBitsPerSecond 500MB`. Transfer was slower now. Challenge times still skyrocketed.

I think I'll try setting up a Pi4 being Full Node and Farmer? I can't see why that would help me with the challenge times **if** it's the NAS' issue though.

**Reply by sargonas** [reply]
Oof... glad to hear it's narrowed down, but bummer it is indeed your NAS. :/ (I, too, hate Windows in scenarios like this... "oh hey I know you set this setting but here's the thing, I have 3 different ways to toggle something from 7 different functions due to 8 layers of legacy code andUIs sooooooo nah I don't think so bud.")

**Reply by Akaino** [reply]
I don't quite get the issue here.
I see how it takes longer to check the plots. However, as long as there are 0 plots passed, everything is quite under 30s.
Once a plot passes, the check takes ages. This can't be the NASs issue alone.

This is now during robocopy moving the files from SSD to NAS...

```
16:30:51.669 harvester chia.harvester.harvester: INFO     1 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 94.62579 s. Total 180 plots
16:32:11.037 harvester chia.harvester.harvester: INFO     1 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 133.66629 s. Total 180 plots
16:33:57.899 harvester chia.harvester.harvester: INFO     1 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 135.03200 s. Total 180 plots
16:33:57.902 harvester chia.harvester.harvester: INFO     2 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 164.53450 s. Total 180 plots
16:33:57.933 harvester chia.harvester.harvester: INFO     1 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 106.88750 s. Total 180 plots
```

```
16:27:31.186 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.42550 s. Total 180 plots
16:27:44.845 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.65800 s. Total 180 plots
16:28:11.816 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.44450 s. Total 180 plots
16:28:53.457 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.48800 s. Total 180 plots
16:29:01.432 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 6.94750 s. Total 180 plots
16:29:17.042 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 10.99400 s. Total 180 plots
16:29:57.370 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.47700 s. Total 180 plots
16:30:24.285 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.54300 s. Total 180 plots
16:30:37.994 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.70800 s. Total 180 plots
16:30:51.660 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.66479 s. Total 180 plots
16:31:06.263 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.57800 s. Total 180 plots
16:31:13.366 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 7.10250 s. Total 180 plots
16:31:29.376 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 11.35900 s. Total 180 plots
16:31:42.866 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.48850 s. Total 180 plots
16:32:10.018 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.53600 s. Total 180 plots
16:32:38.371 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.66400 s. Total 180 plots
16:33:19.482 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 13.52550 s. Total 180 plots
16:33:27.658 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 8.17450 s. Total 180 plots
16:33:32.255 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 4.59650 s. Total 180 plots
16:33:43.357 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming b3a98ece13... Found 0 proofs. Time: 11.10100 s. Total 180 plots
```

**Reply by Akaino** [reply]
For those interested, I mitigated my issue using robocopy with it's _/ipg_ parameter.
This parameter depends on your network connection so I can't really tell you what's best for you... I'm going with this:
`robocopy $source_folder $destination_folder_NAS /MOV *.plot /IPG:10`
Works pretty well so far. No issues.

**Comment by nomadengineer** [comment]
It could be down to Random seek times during a file copy?

**Reply by Akaino** [reply]
I think it was just Windows being Windows... Copy processes seem to be considered highest priority. Everything else went down for the entire duration.
I mitigated it with a robocopy delay though (ipg:10 in _my_ case).
`robocopy $source_folder $destination_folder_NAS /MOV *.plot /IPG:10`

**Reply by nomadengineer** [reply]
I would like to understand this in a bit more detail please.
So when the chia daemon is copying the files from tmpplot -> farm directory, windows being windows will always cause these times to jump?
So we keep the plots in another folder on the tmp drive and use robocopy to send to the farming destination?

I hope I got it!
Thanks.

**Comment by milan-patartics** [comment]
I think challenge processes should run as the highest priority processes by default.

**Comment by BrandtH22** [comment]
If you are still experiencing the issues it is recommended to upgrade to the latest version of chia. Please download it from here: https://www.chia.net/downloads/

Since this issue has been open for a number of years without additional comments, we will be closing this ticket but if we have closed this ticket in error do not hesitate to reach out to us again with any followup questions, comments, or if the issue persists after an update.

```
Please note that upgrading major versions of chia generally requires the config.yaml to be recreated using these steps:
1. Stop chia
2. Rename the config or move it to a different directory
3. In CLI run chia init
4. Copy any custom settings from the old config to the new (plot drive directories, pool info, trusted node info, etc) - be careful here as yaml is very picky about formatting including those leading spaces
5. Start Chia
```

The best place to reach our support team is on Discord (https://discord.gg/chia) or by reopening this ticket.
